WIP: ConvertFrom/To-CliXml

IISResetMe opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

Fix #8311

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

IISResetMe committed:
Move CliXml Writer settings out of ExportCliXmlCommand

IISResetMe committed:
Move ExportCliXmlCommand calls to Helper class

IISResetMe committed:
Add StreamWriter to ExportXmlHelper

IISResetMe committed:
Move file handling calls back to ExportCliXml cmdlet

Decoupling file handling from the helper allows us to re-use it for string conversion

IISResetMe committed:
Add ConvertToClixml cmdlet

Almost same user interface as ConvertToXml, but wrapping the clixml Serializer instead

IISResetMe committed:
Remove using() stmt to avoid race condition in cleanup

IISResetMe committed:
Update Microsoft.PowerShell.Utility exports

IISResetMe committed:
Add ConvertFrom-CliXml

Design to accept both strings and XmlNode input (although ineficiently)

PoshChan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -120,6 +120,8 @@ public SwitchParameter NoClobber
 
```
@IISResetMe, your last commit had 4 failures in `PowerShell-CI-macos`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for collections
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 136
136: 		$out2.Count | Should -Be 2
```

PoshChan replied:
@IISResetMe, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator=<=}, @{InputObject=ConvertTo-Clixml; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 540
540:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator=<=}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 564
564:         $result | Should -BeNullOrEmpty
```
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```

PoshChan replied:
@IISResetMe, your last commit had 4 failures in `PowerShell-CI-linux`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for collections
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 136
136: 		$out2.Count | Should -Be 2
```

PoshChan replied:
@IISResetMe, your last commit had 3 failures in `PowerShell-CI-windows`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator=<=}, @{InputObject=ConvertTo-Clixml; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 540
540:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator=<=}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 564
564:         $result | Should -BeNullOrEmpty
```
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

PoshChan replied:
@IISResetMe, your last commit had 1 failures in `PowerShell-CI-linux`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

PoshChan replied:
@IISResetMe, your last commit had 1 failures in `PowerShell-CI-macos`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

PoshChan replied:
@IISResetMe, your last commit had 2 failures in `PowerShell-CI-macos`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

PoshChan replied:
@IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

PoshChan replied:
@IISResetMe, your last commit had 2 failures in `PowerShell-CI-macos`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

PoshChan replied:
@IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

PoshChan replied:
@IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

IISResetMe committed:
Add XmlNode support, restore Paging support in Import-CliXml

Previous commits removed implementation of paging in Import-CliXml, fixed.
Added ImportXmlHelper overloader that takes a reader instead of a stream, to avoid converting already parsed Xml back to strings

IISResetMe committed:
Add ConvertTo/From-Clixml to list of default commands to test for

IISResetMe committed:
Fix Bug161470 test

It's unclear what this specific (regression?) test is designed for, but if it's expected to throw an error then we should probably test for that?

IISResetMe committed:
Fix ProcessXmlNode codepath, cleanup CodeFactor findings

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -362,6 +316,236 @@ protected override void StopProcessing()
+        /// </summary>
+        protected override void ProcessRecord()
+        {
+            if (_serializer != null)
```
It looks like _serializer is always created in 'CreateMemoryStream', so there is no need to check if it is null.

IISResetMe replied:
Yeah, you're right, but it should also be moved completely back to the `ExportXmlHelper` class. 

The idea here is that `ConvertTo/Export-Clixml` and `ConvertFrom/Import-CliXml` become thin wrappers around the helpers which in turn abstracts away the the (de-)serialization interfaces, and the cmdlets themselves just feed a reader or writer to the helper class.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -362,6 +316,236 @@ protected override void StopProcessing()
+        /// </summary>
+        protected override void EndProcessing()
+        {
+            if (_serializer != null)
```
Null check appears unnecessary.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -623,38 +807,157 @@ private void CleanUp()
+
+        internal void Serialize(object inputObject)
+        {
+            if (_serializer != null)
```
Looks like _serializer is created in the constructor, so don't need null check.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -679,18 +982,24 @@ private static XmlReader CreateXmlReader(Stream stream)
             return XmlReader.Create(textReader, InternalDeserializer.XmlReaderSettingsForCliXml);
```
This seems wrong as a new reader is created every time the property is accessed.

IISResetMe committed:
Fix inappropriate cleanup

Disposing of the Stream instances injected into Helper classes should be the responsibility of the caller, not the Helper

IISResetMe committed:
Remove unnecessary null checks

IISResetMe committed:
Clean up Export-CliXml

IISResetMe committed:
Add ConvertTo/From-CliXml to powershell-unix
